flamingpy/noise/cv.py

```diff
@@ -0,0 +1,649 @@
+            covs = cov_dict[quad][inds, :][:, inds].toarray()
+            outcomes = rng.multivariate_normal(mean=means, cov=covs, method=method)
+        for i in range(N_inds):
+            self.egraph.nodes[self.to_points[inds[i]]]["hom_val_" + quad] = outcomes[i]
+
+    def SCZ(self, sparse=False):
+        """Return the symplectic matrix associated with CZ application.
+
+        Returns:
+            array: the symplectic matrix.
+        """
+        adj = self._adj
+        return SCZ_mat(adj, sparse)
+
+    def hom_outcomes(self, inds=None, quad="p"):
+        """array: quad-homodyne measurement outcomes for modes inds."""
+        N = self._N
+        if inds is None:
+            inds = range(N)
+        outcomes = [self.egraph.nodes[self.to_points[i]].get("hom_val_" + quad) for i in inds]
+        return outcomes
+
+    def bit_values(self, inds=None):
+        """array: bit values associated with the p measurement."""
+        N = self._N
+        if inds is None:
+            inds = range(N)
+        bits = [self.egraph.nodes[self.to_points[i]].get("bit_val") for i in inds]
+        return bits
+
+    @property
+    def p_inds(self):
+        """array: the indices of the p-squeezed states."""
+        return self._states.get("p")
+
+    @property
+    def GKP_inds(self):
+        """array: the indices of the GKP states."""
+        return self._states.get("GKP")
+
+    def draw(self, **kwargs):
+        """Draw the CV graph state with matplotlib.
+
+        See flamingpy.utils.viz.draw_EGraph for more details. Use the default
+        colours: gold for GKP states and blue for p-squeezed
+        states.
+        """
+        cv_opts = {"color_nodes": ("state", {"GKP": "gold", "p": "blue"})}
+        updated_opts = {**cv_opts, **kwargs}
+        return self.egraph.draw(**updated_opts)
+
+    def draw_SCZ(self, **kwargs):
+        """Draw the adjacency matrix of a CV graph state with matplotlib.
+
+        See flamingpy.utils.viz.plot_mat_heat_map for more details.
+        """
+        from flamingpy.utils.viz import plot_mat_heat_map
+
+        return plot_mat_heat_map(self.SCZ(), **kwargs)
+
+
+four_splitter = splitter_symp()
+
+
+class CVMacroLayer(CVLayer):
+    """A class for reducing a macronode CV graph to a canonical graph.
+
+    Applies noise to self.egraph (assumed a macronode graph), entangles the
+    macronodes, measures the syndrome, and populates the canonical graph
+    reduced_graph with the reduced states, bit values, and error probabilities.
+
+    In addition to CVLayer args, the following --
+
+    Args:
+        reduced_graph (EGraph): the canonical (reduced) code graph.
+        bs_network (np.array, optional): the sympletic matrix corresponding to
+            the beamsplitter network entangling the macronode. By default,
+            the standard four-splitter.
+    """
+
+    def __init__(self, *args, reduced_graph, bs_network=None, **kwargs):
+        super().__init__(*args, **kwargs)
```

Comment:
There are a couple of issues with this.

- CVLayer applies cv operations on `code` (if an EGraph) and `code.graph` (otherwise). We would like CVMacroLayer (which inherits from CVLayer) to apply cv ops on the macronized graph, not the EGraph. 
- We do not want to regenerate the macronized graph every Monte Carlo trial.

Reply:
Why do you need to rebuild CVMacroLayer everytime? This is not clear to me.

If you have a good reason to do so, then forget about calling macronize in the init. However, the `*args` part still need to go. Just input the graph directly.